File: src/func/solve.ts
import Trie from "./Trie";

export function solve_bfs(nums: number[][], n: number, correct_board: number[][]) {
  const visited = new Trie();
  const path = new Map();

  path.set(nums, "");
  const to_visit = [nums];
  while (to_visit.length != 0) {
    const current = to_visit.pop();
    if (!current) {
      break;
    }
    const moves = find_move_opportunity(current, n);
    for (let index = 0; index < moves.length; index++) {
      const move = moves[index];
      if (check_win(move, correct_board)) {
        let curr = current;
        const res_path = [move];
        while (curr) {
          res_path.push(curr);
          curr = path.get(curr);
        }
        return res_path.reverse();
      }
      if (visited.exists(move)) {
        continue;
      }
      visited.add(move);
      path.set(move, current);
      to_visit.unshift(move);
    }
  }
  console.log("sad");
}

export function solve_dfs(nums: number[][], root: Trie): number[][][] {
  if (root.exists(nums)) {
    return [];
  }
  if (check_win(nums, [[]])) {
    return [nums];
  }
  root.add(nums);
  const moves = find_move_opportunity(nums, 0);
  for (let index = 0; index < moves.length; index++) {
    const element = moves[index];
    const result = solve_dfs(element, root);
    if (result.length > 0) {
      result.push(nums);
      return result;
    }
  }
  return [];
}

function find_zero(nums: number[][], n: number) {
  for (let i = 0; n > i; i++) {
    for (let j = 0; j < n; j++) {
      const element = nums[i][j];
      if (element == 0) {
        return [i, j];
      }
    }
  }
  return [0, 0];
}

function find_move_opportunity(nums: number[][], n: number) {
  const zero = find_zero(nums, n);
  const i = zero[0];
  const j = zero[1];
  const next_index_list = [];

  if (i > 0) {
    let num_mod = nums.map((row) => row.slice());
    const last_num = num_mod[i - 1][j];
    num_mod[i][j] = last_num;
    num_mod[i - 1][j] = 0;
    next_index_list.push(num_mod);
  }

  if (i < n - 1) {
    let num_mod = nums.map((row) => row.slice());
    const last_num = num_mod[i + 1][j];
    num_mod[i][j] = last_num;
    num_mod[i + 1][j] = 0;
    next_index_list.push(num_mod);
  }

  if (j > 0) {
    let num_mod = nums.map((row) => row.slice());
    const last_num = num_mod[i][j - 1];
    num_mod[i][j] = last_num;
    num_mod[i][j - 1] = 0;
    next_index_list.push(num_mod);
  }

  if (j < n - 1) {
    let num_mod = nums.map((row) => row.slice());
    const last_num = num_mod[i][j + 1];
    num_mod[i][j] = last_num;
    num_mod[i][j + 1] = 0;
    next_index_list.push(num_mod);
  }
  return next_index_list;
}

export function check_win(nums: number[][], correct_board: number[][]) {
  for (let i = 0; i < nums.length; i++) {
    for (let j = 0; j < nums[i].length; j++) {
      if (nums[i][j] != correct_board[i][j]) {
        return false;
      }
    }
  }
  return true;
}

export function createCorrectBoard(n: number) {
  const nums: number[][] = [];
  let i = -1;
  for (let index = 0; index < n * n - 1; index++) {
    if (index % n == 0) {
      i++;
      nums.push([index + 1]);
    } else {
      try {
        nums[i].push(index + 1);
      } catch {
        console.log(i);
        console.log(nums);
      }
    }
  }
  nums[nums.length - 1].push(0);
  console.log("board_correct");
  console.log(nums);
  return nums;
}

function log_nums(nums: number[][], i: number, j: number) {
  console.log("------------------------------------------------");
  console.log(`i: ${i}, j: ${j}`);
  nums.forEach((element) => {
    console.log(element);
  });
  console.log("------------------------------------------------");
}

export function move(nums_p: number[][], i: number, j: number, n: number) {
  let nums = nums_p.map((row) => row.slice());
  const zeroPosition = find_zero(nums, n);
  if (
    (i - 1 <= zeroPosition[0] && i + 1 >= zeroPosition[0] && zeroPosition[1] == j) ||
    (j - 1 <= zeroPosition[1] && j + 1 >= zeroPosition[1] && zeroPosition[0] == i)
  ) {
    const last_num = nums[i][j];
    nums[i][j] = 0;
    nums[zeroPosition[0]][zeroPosition[1]] = last_num;
    return nums;
  }
}

File: internal/solve.go
package internal

import (
	"container/list"
	"fmt"
)

const Size_8 int8 = 4
const Size = 4

type Data struct {
	data   [Size][Size]int8
	move   int8
	parent *Data
}

func Solve_bfs(nums [Size][Size]int8) [][Size][Size]int8 {
	visited := make(map[[Size][Size]int8]struct{})

	to_visit := list.New()
	to_visit.PushBack(Data{data: nums, move: 0, parent: nil})
	correctBoard := createCorrectBoard()
	for to_visit.Len() != 0 {
		current_element := to_visit.Front()
		to_visit.Remove(current_element)
		current := current_element.Value.(Data)
		moves := findMoveOpportunity(current.data)

		for _, v := range moves {
			if v == correctBoard {
				var result [][Size][Size]int8
				for {
					result = append(result, current.data)
					if current.parent == nil {
						return result
					}
					current = *current.parent
				}
			}
			_, exists := visited[v]
			if exists {
				continue
			}
			visited[v] = struct{}{}
			to_visit.PushBack(Data{data: v, move: 0, parent: &current})
		}
	}
	return nil
}

func createCorrectBoard() [Size][Size]int8 {
	var nums [Size][Size]int8
	for index := 0; index < Size*Size-1; index++ {
		nums[index/Size][index%Size] = int8(index + 1)
	}

	// Add the 0 to the last row
	nums[Size-1][Size-1] = 0

	fmt.Println("board_correct")
	fmt.Println(nums)

	return nums
}

// findZero finds the position of 0 in the matrix
func findZero(nums [Size][Size]int8) (int8, int8) {
	for i := 0; i < Size; i++ {
		for j := 0; j < Size; j++ {
			if nums[i][j] == 0 {
				return int8(i), int8(j)
			}
		}
	}
	return 0, 0
}

// deepCopyMatrix creates a deep copy of a 2D slice
func deepCopyMatrix(nums [Size][Size]int8) [Size][Size]int8 {
	return nums
}

// findMoveOpportunity returns all possible boards after moving the zero
func findMoveOpportunity(nums [Size][Size]int8) [][Size][Size]int8 {
	i, j := findZero(nums)
	nextIndexList := make([][Size][Size]int8, 4)

	// move up
	if i > 0 {
		numMod := deepCopyMatrix(nums)
		numMod[i][j], numMod[i-1][j] = numMod[i-1][j], 0
		nextIndexList = append(nextIndexList, numMod)
	}

	// move down
	if i < Size_8-1 {
		numMod := deepCopyMatrix(nums)
		numMod[i][j], numMod[i+1][j] = numMod[i+1][j], 0
		nextIndexList = append(nextIndexList, numMod)
	}

	// move left
	if j > 0 {
		numMod := deepCopyMatrix(nums)
		numMod[i][j], numMod[i][j-1] = numMod[i][j-1], 0
		nextIndexList = append(nextIndexList, numMod)
	}

	// move right
	if j < Size_8-1 {
		numMod := deepCopyMatrix(nums)
		numMod[i][j], numMod[i][j+1] = numMod[i][j+1], 0
		nextIndexList = append(nextIndexList, numMod)
	}

	return nextIndexList
}
